File: 3_longest_substring/3_longest_substring.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;
        
        int bestScore = 0;
        Map<Character, Integer> letters = new HashMap<>();
        int start = 0;
        int end;

        for (end = 0; end < s.length(); end++) {
            char c = s.charAt(end);
            System.out.println(end + " " + c + ", " + start + " " + s.charAt(start) + ": " + letters);

            if (!letters.containsKey(c) || letters.get(c) == -1) {
                letters.put(c, end);
                continue;
            }

            if (bestScore < end - start) {
                bestScore = end - start;
            }
            
            int previous = letters.get(c);

            for (int i = start; i <= previous; i++) {
                letters.put(s.charAt(i), -1);
            }

            start = previous + 1;
            letters.put(s.charAt(start), start);
            letters.put(c, end);
        }

        if (bestScore < end - start) {
            bestScore = end - start;
        }

        return bestScore;
    }
}

File: 3_longest_substring/3_longest_substring.py
import time


def longestSubstring(s):
    if len(s) == 0: 
        return 0
    
    bestScore = 0
    letters = {}
    start = 0

    for end in range(len(s)):
        previous = letters.get(s[end])
        if previous == None or previous == -1:
            letters[s[end]] = end
            continue

        if bestScore < end - start:
            bestScore = end - start

        for i in range(start, previous + 1):
            letters[s[i]] = -1

        start = previous + 1
        letters[s[start]] = start
        letters[s[end]] = end

    if bestScore < end+1 - start:
        bestScore = end+1 - start 

    return bestScore